cmd/all-in-one/main.go

```diff
@@ -317,32 +315,25 @@ func startQuery(
 		logger.Fatal("Failed to initialize tracer", zap.Error(err))
 	}
 	opentracing.SetGlobalTracer(tracer)
+	return closer
+}
 
-	spanReader = storageMetrics.NewReadMetricsDecorator(spanReader, baseFactory.Namespace(metrics.NSOptions{Name: "query", Tags: nil}))
-
-	qs := querysvc.NewQueryService(spanReader, depReader, queryOpts)
-	handlerOpts := []queryApp.HandlerOption{queryApp.HandlerOptions.Logger(logger), queryApp.HandlerOptions.Tracer(tracer)}
-	apiHandler := queryApp.NewAPIHandler(
-		qs,
-		handlerOpts...)
-
-	r := mux.NewRouter()
-	if qOpts.BasePath != "/" {
-		r = r.PathPrefix(qOpts.BasePath).Subrouter()
+func startQuery(
+	svc *flags.Service,
+	qOpts *queryApp.QueryOptions,
+	queryOpts *querysvc.QueryServiceOptions,
+	spanReader spanstore.Reader,
+	depReader dependencystore.Reader,
+	rootFactory metrics.Factory,
+	baseFactory metrics.Factory,
+) *queryApp.Server {
+	spanReader = storageMetrics.NewReadMetricsDecorator(spanReader, baseFactory.Namespace(metrics.NSOptions{Name: "query"}))
+	qs := querysvc.NewQueryService(spanReader, depReader, *queryOpts)
+	server := queryApp.NewServer(svc, qs, qOpts, opentracing.GlobalTracer())
+	if err := server.Start(); err != nil {
+		svc.Logger.Fatal("Could not start jaeger-query service", zap.Error(err))
```

Comment:
If the server fail to start, no healthcheck update, how do we know the query service failed beside checking the log?

Reply:
Valid question, but not really related to this PR - we'd need to redesign how we handle failures to start in the servers that run in background goroutines. https://github.com/jaegertracing/jaeger/issues/1508

Reply:
OK